crates/scmi/src/devices/iio.rs

```diff
@@ -0,0 +1,826 @@
+// SPDX-FileCopyrightText: Red Hat, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Industrial I/O (IIO) sensors bindings.
+//!
+//! Basic functionality for exposing `/sys/bus/iio/devices/` stuff as guest
+//! SCMI devices.  Only some typical cases are supported.  If you want more
+//! functionality, you must enhance the implementation here.
+//!
+//! For some entry points, see [IIOSensor] and [Axis].
+
+use std::cmp::{max, min};
+use std::ffi::{OsStr, OsString};
+use std::fs;
+use std::io::ErrorKind;
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+
+use log::{debug, error, warn};
+
+use crate::scmi::{self, DeviceResult, MessageValue, ScmiDeviceError, MAX_SIMPLE_STRING_LENGTH};
+
+use super::common::{
+    check_device_properties, DeviceError, DeviceProperties, MaybeDevice, Sensor, SensorDevice,
+    SensorT,
+};
+
+/// Information about units used by the given Linux IIO channel.
+struct UnitMapping<'a> {
+    /// IIO sysfs channel prefix, e.g. "in_accel".
+    channel: &'a str,
+    /// One of the SCMI unit constants from [crate::scmi] (enum is not used to
+    /// avoid type conversions everywhere).
+    unit: u8,
+    /// Decadic exponent to be used to convert the given unit to the SCMI unit.
+    /// For example, the exponent is 0 for no conversion, -3 to convert
+    /// milliamps here to amps in SCMI, or 3 to convert kilopascals here to
+    /// pascals in SCMI.
+    unit_exponent: i8, // max. 5 bits actually
+}
+
+/// Specification of IIO channel units.
+///
+/// Based on
+/// <https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/ABI/testing/sysfs-bus-iio>.
+/// Not everything from there is present -- channels here with more complicated
+/// unit transformations (beyond using a decadic exponent; e.g. degrees to
+/// radians or units not defined in SCMI) are omitted.  If an IIO channel
+/// doesn't have unit specification here, it can be still used by the unit
+/// reported in SCMI will be [crate::scmi::SENSOR_UNIT_UNSPECIFIED].
+// TODO: Make some macro(s) for this?
+const UNIT_MAPPING: &[UnitMapping] = &[
+    UnitMapping {
+        channel: "in_accel",
+        unit: scmi::SENSOR_UNIT_METERS_PER_SECOND_SQUARED,
+        unit_exponent: 0,
+    },
+    UnitMapping {
+        channel: "in_angle",
+        unit: scmi::SENSOR_UNIT_RADIANS,
```

Comment:
Suggestion: You could use https://github.com/rust-phf/rust-phf for this (not in this PR necessarily, but in general)

Reply:
Thanks for the tip, good to know.  (In this PR, I don't think there is a good enough reason to introduce another dependency for the given purpose.)

Reply:
yeah it was just a thought :D